### PR TITLE
MINOR: improve RecordCollectorImpl

### PR DESCRIPTION
**Pull request by mjsax**
Found this while working on https://github.com/apache/kafka/pull/17169, but did not want to piggy-back this change to the other PR but isolate as it's unrelated cleanup.

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java [hunk 1]
```diff
@@ -81,7 +81,7 @@ public class RecordCollectorImpl implements RecordCollector {
     private final Sensor droppedRecordsSensor;
     private final Map<String, Sensor> producedSensorByTopic = new HashMap<>();
 
-    private final AtomicReference<KafkaException> sendException;
+    private final AtomicReference<KafkaException> sendException = new AtomicReference<>(null);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java]: There is no reason to get this from `StreamsProducer` so pulling it into this class.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java [hunk 2]
```diff
@@ -529,7 +528,7 @@ private void checkForException() {
         final KafkaException exception = sendException.get();
 
         if (exception != null) {
-            sendException.set(null);
+            sendException.compareAndSet(exception, null);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java:558]: I believe this is actually fixing a race condition... with `set()` we might overwrite (and thus drop) and exception which is set after we called `get()` (even if I am not 100% sure how bad it is if we drop it, give that we `throw exception;` below anyway...?

Comment (lucasbru) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java:558]: Yeah... not sure that part makes a difference. But I guess there is nothing wrong with it?

The rest of PR makes sense to me

**Review by lucasbru — APPROVED** [review]
LGTM, thanks!

**Comment by mjsax** [comment]
After re-thinking, the test we remove because it failed actually exposed that the original idea of this PR is incorrect. We need to keep `sendException` inside `StreamsProducer` as we need a single instance per thread, shared across all tasks (ie, all `RecordCollectorImpl` -- we have one per task).

**Review by lucasbru — APPROVED** [review]

**Pull request merged**